crates/polars-lazy/src/physical_plan/executors/scan/csv.rs

```diff
@@ -35,21 +38,143 @@ impl CsvExec {
-            .with_row_index(self.file_options.row_index.clone())
-            .with_path(Some(self.path.clone()))
-            .try_into_reader_with_file_path(None)?
-            ._with_predicate(predicate)
-            .finish()
+            .with_rechunk(
+                // We rechunk at the end to avoid rechunking multiple times in the
+                // case of reading multiple files.
+                false,
+            )
+            .with_row_index(None)
+            .with_path::<&str>(None);
+
+        let verbose = verbose();
+
+        let mut df =
+            if n_rows.is_some() || (predicate.is_some() && self.file_options.row_index.is_some()) {
+                // Basic sequential read
+                // predicate must be done after n_rows and row_index, so we must read sequentially
+                if verbose {
+                    eprintln!("read per-file to apply n_rows or (predicate + row_index)");
+                }
+
+                let mut n_rows_read = 0usize;
+                let mut out = Vec::with_capacity(self.paths.len());
+                // If we have n_rows or row_index then we need to count how many rows we read, so we need
+                // to delay applying the predicate.
+                let predicate_during_read = predicate
+                    .clone()
+                    .filter(|_| n_rows.is_none() && self.file_options.row_index.is_none());
+
+                for i in 0..self.paths.len() {
+                    let path = &self.paths[i];
+
+                    let df = options_base
+                        .clone()
+                        .with_row_index(self.file_options.row_index.clone().map(|mut ri| {
+                            ri.offset += n_rows_read as IdxSize;
+                            ri
+                        }))
+                        .with_n_rows(n_rows.map(|n| n - n_rows_read))
+                        .try_into_reader_with_file_path(Some(path.clone()))
+                        .unwrap()
+                        ._with_predicate(predicate_during_read.clone())
+                        .finish()?;
+
+                    n_rows_read = n_rows_read.saturating_add(df.height());
+
+                    let df = if predicate.is_some() && predicate_during_read.is_none() {
+                        let predicate = predicate.clone().unwrap();
+
+                        concat_df(
+                            POOL.install(|| {
+                                df.split_chunks()
+                                    .collect::<Vec<_>>()
+                                    .into_par_iter()
+                                    .map(|df| {
+                                        let s = predicate.evaluate_io(&df)?;
+                                        let mask = s
+                                            .bool()
+                                            .expect("filter predicates was not of type boolean");
+                                        df._filter_seq(mask)
```

Comment:
I think the parallel filter might be better here. There could be cases where we have more threads than chunks. Or there can be cases where some columns are more expensive to filter than others (string columns, lists and structs), so having more tasks to share can be beneficial.